pkg/validation/usage_groups.go

```diff
@@ -0,0 +1,187 @@
+			Help:      "The total number of decompressed bytes per profile received by usage group.",
+		},
+		[]string{"type", "tenant", "usage_group"},
+	)
+
+	// This is a duplicate of discarded_bytes_total, but with usage_group as a
+	// label.
+	usageGroupDiscardedBytes = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "pyroscope",
+			Name:      "usage_group_discarded_bytes_total",
+			Help:      "The total number of bytes that were discarded by usage group.",
+		},
+		[]string{"reason", "tenant", "usage_group"},
+	)
+)
+
+type UsageGroupConfig struct {
+	TenantID string
+	config   map[string][]*labels.Matcher
+}
+
+func (c *UsageGroupConfig) GetUsageGroups(lbls phlaremodel.Labels) UsageGroupMatch {
+	match := UsageGroupMatch{
+		tenantID: c.TenantID,
+	}
+
+	for name, matchers := range c.config {
+		if matchesAll(matchers, lbls) {
+			match.names = append(match.names, name)
+		}
+	}
+
+	return match
+}
+
+func (c *UsageGroupConfig) UnmarshalYAML(value *yaml.Node) error {
+	m := make(map[string]string)
+	err := value.DecodeWithOptions(&m, yaml.DecodeOptions{
+		KnownFields: true,
+	})
+	if err != nil {
+		return fmt.Errorf("malformed usage group config: %w", err)
+	}
+
+	*c, err = NewUsageGroupConfig(m)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *UsageGroupConfig) UnmarshalJSON(bytes []byte) error {
+	m := make(map[string]string)
+	err := json.Unmarshal(bytes, &m)
+	if err != nil {
+		return fmt.Errorf("malformed usage group config: %w", err)
+	}
+
+	*c, err = NewUsageGroupConfig(m)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type UsageGroupMatch struct {
+	tenantID string
+	names    []string
+}
+
+func (m UsageGroupMatch) CountReceivedBytes(profileType string, n int64) {
+	if len(m.names) == 0 {
+		usageGroupReceivedDecompressedBytes.WithLabelValues(profileType, m.tenantID, noMatchName).Add(float64(n))
+		return
+	}
+
+	for _, name := range m.names {
+		usageGroupReceivedDecompressedBytes.WithLabelValues(profileType, m.tenantID, name).Add(float64(n))
+	}
+}
+
+func (m UsageGroupMatch) CountDiscardedBytes(reason string, n int64) {
+	if len(m.names) == 0 {
+		usageGroupDiscardedBytes.WithLabelValues(reason, m.tenantID, noMatchName).Add(float64(n))
+		return
+	}
+
+	for _, name := range m.names {
+		usageGroupDiscardedBytes.WithLabelValues(reason, m.tenantID, name).Add(float64(n))
+	}
+}
+
+func NewUsageGroupConfig(m map[string]string) (UsageGroupConfig, error) {
+	if len(m) > maxUsageGroups {
+		return UsageGroupConfig{}, fmt.Errorf("maximum number of usage groups is %d, got %d", maxUsageGroups, len(m))
+	}
+
+	config := UsageGroupConfig{
+		config: make(map[string][]*labels.Matcher),
+	}
+
+	for name, matchersText := range m {
+		name = strings.TrimSpace(name)
+		if name == "" {
+			return UsageGroupConfig{}, fmt.Errorf("usage group name cannot be empty")
+		}
+
+		// TODO(bryanhuhta): We should probably validate the usage group name
+		// is a valid label value.
```

Comment:
I couldn't figure out where prometheus does its validation of label values, but wherever it is, we should copy it.

Reply:
It is any valid utf8 string: https://github.com/prometheus/client_golang/blob/9f203a098ec630d179ccef4efbaa8c3341291d00/prometheus/labels.go#L158


I suggest we also disallow "other" to avoid undefined behaviour.